servicetalk-log4j2-mdc/src/main/java/io/servicetalk/log4j2/mdc/MdcCapturedContextProvider.java

```diff
@@ -0,0 +1,109 @@
+ * Note: this class should not be used directly: it is only intended to be service loaded by ServiceTalk.
+ */
+public final class MdcCapturedContextProvider implements CapturedContextProvider {
+
+    private final boolean enabled;
+
+    /**
+     * Create a new {@link CapturedContextProvider} for MDC.
+     * Note: this class should not be used directly: it is only intended to be service loaded by ServiceTalk.
+     */
+    public MdcCapturedContextProvider() {
+        enabled = shouldEnableMdcCapture();
+    }
+
+    @Override
+    public CapturedContext captureContext(CapturedContext underlying) {
+        return enabled ? new MdcCapturedContext(underlying, getCurrent()) : underlying;
+    }
+
+    @Override
+    public CapturedContext captureContextCopy(CapturedContext underlying) {
+        return enabled ? new MdcCapturedContext(underlying, getCurrentCopy()) : underlying;
+    }
+
+    private static final class MdcCapturedContext implements CapturedContext {
+
+        private final CapturedContext delegate;
+        private final Map<String, String> storage;
+
+        MdcCapturedContext(CapturedContext delegate, Map<String, String> storage) {
+            this.delegate = delegate;
+            this.storage = storage;
+        }
+
+        @Override
+        public ContextMap captured() {
+            return delegate.captured();
+        }
+
+        @Override
+        public Scope attachContext() {
+            Map<String, String> old = getCurrent();
+            setCurrent(storage);
+            Scope delegateScope = delegate.attachContext();
+            return () -> {
+                delegateScope.close();
+                setCurrent(old);
+            };
+        }
+    }
+
+    private static Map<String, String> getCurrent() {
+        return DefaultServiceTalkThreadContextMap.CONTEXT_STORAGE.get();
+    }
+
+    private static Map<String, String> getCurrentCopy() {
+        Map<String, String> current = DefaultServiceTalkThreadContextMap.CONTEXT_STORAGE.get();
+        Map<String, String> result = new ConcurrentHashMap<>(Math.max(current.size(), 4));
+        result.putAll(current);
+        return result;
+    }
+
+    private static void setCurrent(Map<String, String> storage) {
+        DefaultServiceTalkThreadContextMap.CONTEXT_STORAGE.set(storage);
+    }
+
+    // TODO: we could attempt a 'slow path' where we can copy and propagate values over whatever MDC context provider
+    //  is present, but for now we're just going to disable it since that would be what happens anyway.
+    @SuppressWarnings({"UseOfSystemOutOrSystemErr", "PMD.SystemPrintln"})
+    private static boolean shouldEnableMdcCapture() {
+        ReadOnlyThreadContextMap implementation = ThreadContext.getThreadContextMap();
+        if (implementation instanceof DefaultServiceTalkThreadContextMap) {
+            return true;
+        }
+        System.err.println("Incompatible MDC ThreadContext adapter detected (" +
```

Comment:
can slf4j still be used here even if the thread context map is different? or is this a chicken and egg problem where this is done during log init so it would be swallowed.

Reply:
I don't know for sure, but I strongly suspect a checken-egg problem. This is the same pattern taken in the mdc-utils package, presumably for the same reason.

Reply:
If we have both `DefaultServiceTalkThreadContextMap` and this class in the same package, what could be a reason for this to happen? I assume both should be wired together.

Reply:
We could potentially have conflicting MDC ThreadContextMap Providers. This plays to the TODO above about having a slow-path for that situation, but it could end up being more harm than good.